src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -0,0 +1,270 @@
+        Raises:
+          AioRpcError: Indicating that the RPC terminated with non-OK status.
+          asyncio.CancelledError: Indicating that the RPC was canceled.
+        """
+
+
+class InterceptedUnaryUnaryCall(_base_call.UnaryUnaryCall):
+    """Used for running a `UnaryUnaryCall` wrapped by interceptors, where
+    interceptors might have some work to do before the RPC invocation with
+    the capacity of changing the invocation parameters, and some work to do
+    after the RPC invocation with the capacity for accessing to the wrapped
+    `UnaryUnaryCall`.
+
+    It handles also early and later cancellations, when the RPC has not even
+    started and the execution is still held by the interceptors or when the
+    RPC has finished but again the execution is still held by the interceptors.
+
+    Once the RPC is finally executed, all methods are finally done against the
+    intercepted call, being at the same time the same call returned to the
+    interceptors.
+
+    For most of the methods, like `initial_metadata()` the caller does not need
+    to wait until the interceptors task is finished, once the RPC is done the
+    caller will have the freedom for accessing to the results.
+
+    For the `__await__` method is it is proxied to the intercepted call only when
+    the interceptor task is finished.
+    """
+
+    _loop: asyncio.AbstractEventLoop
+    _channel: cygrpc.AioChannel
+    _cancelled_before_rpc: bool
+    _intercepted_call: Optional[_base_call.UnaryUnaryCall]
+    _intercepted_call_created: asyncio.Event
+    _interceptors_task: asyncio.Task
+
+    def __init__(
+            self,
+            interceptors: Sequence[UnaryUnaryClientInterceptor],  # pylint: disable=R0913
+            request: RequestType,
+            timeout: Optional[float],
+            channel: cygrpc.AioChannel,
+            method: bytes,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> None:
+        self._channel = channel
+        self._loop = asyncio.get_event_loop()
+        self._intercepted_call = None
+        self._intercepted_call_created = asyncio.Event(loop=self._loop)
+        self._cancelled_before_rpc = False
+        self._invoke(interceptors, method, timeout, request, request_serializer,
+                     response_deserializer)
+
+    def __del__(self):
+        self.cancel()
+
+    def _invoke(self, interceptors: Sequence[UnaryUnaryClientInterceptor],
+                method: bytes, timeout: Optional[float], request: RequestType,
+                request_serializer: SerializingFunction,
+                response_deserializer: DeserializingFunction) -> None:
+        """Run the RPC call wrapped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryUnaryClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request: RequestType) -> _base_call.UnaryUnaryCall:
```

Comment:
This is IMO related to the conversation that opened @gnossen here [1].

I guess that I didn't express myself with enough clarity in the previous PR, but I would advocate for not allowing the caller for accessing to the RPC till the interceptor task is not finished. Without this address, the problem that you are presenting can not be solved in a consistent way, meaning that the caller might use a different call than the one returned by the last interceptor.

We would need to make a call on that, as I said I'm in favor of not plumbing the caller to the last call returned by the last interceptor.

Once this is solved we can start talking about the mutation problem, that you are rising here. Technically, once the previous issue is solved, the interceptors would be free to return any call, for example, calls that have been cached previously. This would solve the `CacheInterceptor` example that you provided.

The "problem" would come if we need more granularity, for example for mutating the metadata or changing the status code, which right now is not supported by the `Call` interface. TBH I'm wondering if this is really supported by the sync version. is this supported?

Most likely I'm missing something, but having the feeling that for me the other examples that you have provided - the one for suppressing errors and the one for making a result composition - the way to go would be implementing that specific business logic in the application layer rather than trying to do it using an interceptor. 

[1] https://github.com/grpc/grpc/pull/21455/files#r356895373

Reply:
> The "problem" would come if we need more granularity, for example for mutating the metadata or changing the status code, which right now is not supported by the Call interface. TBH I'm wondering if this is really supported by the sync version. is this supported?

For the sync version, applications can mutate the sending initial metadata by changing the `client_call_details`. As for trailing metadata, you are right, there is no way to mutate it.

> Technically, once the previous issue is solved, the interceptors would be free to return any call, for example, calls that have been cached previously.

To make caching data structure efficient, applications might want to store the data itself without the abstraction. Can you try to write a short snippet for a simple `CacheInterceptor`? Or add it as a unit test.

> ...the problem that you are presenting can not be solved in a consistent way, meaning that the caller might use a different call than the one returned by the last interceptor.

If we can solve the common use cases, I'm fine with either design, in fact the one call design is simpler than the multiple call design.

Reply:
I found there isn't an example for cached interceptor. I wrote my short version:

```Python
import grpc
import helloworld_pb2
import helloworld_pb2_grpc
import logging

class CacheInterceptor(grpc.UnaryUnaryClientInterceptor):

    def __init__(self):
        self._cached = {}

    def intercept_unary_unary(self,
                              continuation,
                              client_call_details,
                              request):
        if client_call_details.method != '/helloworld.Greeter/SayHello':
            # Not-cacheable methods, just proceed.
            return continuation(client_call_details, request)
        hash_key = request.name
        if hash_key in self._cached:
            logging.debug('Cache hit for request [%s]', request)
            return self._cached[hash_key]
        else:
            logging.debug('Cache miss for request [%s]', request)
            response = continuation(client_call_details, request)
            self._cached[hash_key] = response
            return response


def main():
    channel = grpc.insecure_channel('localhost:50051')
    intercepted_channel = grpc.intercept_channel(channel, CacheInterceptor())

    stub = helloworld_pb2_grpc.GreeterStub(intercepted_channel)

    stub.SayHello(helloworld_pb2.HelloRequest(name='First'))
    stub.SayHello(helloworld_pb2.HelloRequest(name='First'))
    stub.SayHello(helloworld_pb2.HelloRequest(name='First'))

    stub.SayHello(helloworld_pb2.HelloRequest(name='Second'))

    stub.SayHello(helloworld_pb2.HelloRequest(name='Third'))

    channel.close()

if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG)
    main()
```

Reply:
Oks, let try to summarize the key points of the talk that we had "off linee" regarding the changes that would need to be done in this PR 

1 - Do not allow the caller to access before the code interceptors are finished to the RPC attributes, the caller would need to see consistently the result of the last call returned by the last interceptor (well indeed is the first one :))

2 - Considering that the synchronous client allows the interceptors to return a raw response, rather than returning always a call. The interceptor interface would give that freedom in for the aio interface. If a response - proto - is returned instead of a `UnaryUnaryCall`, behind the scenes and before returning the value to the prev interceptor the response will be wrapped into a new `UnaryUnaryCall` with default values.

3 - In some scenarios, the interceptor could be interested in just mutating the response returned by a real RPC. For that, we could provide a new method like `call.with_response(proto)` for building a new `UnaryUnaryCall` that would inherit all of the RPC attributes by producing a new call instance where the response would be replaced for the one given as a parameter. 

Option 1 and 2 would be a must for having feature parity with the sync interceptors, while Option 3 could be a nice to have that could be implemented in the future if the users ask for it.  

If nobody opposes I will be doing the proper changes in this PR for implementing option 1 and option 2. /cc @lidizheng @gnossen

Reply:
I'm not fully understanding Option 1, can you explain more?

Option 2 sounds good.

About Option 3, `call.with_response(proto)` is not going to work well with streaming RPCs. Let's rule it out.

Reply:
Option 1 basically means remove support for this [1], so the caller yes or yes would need to wait always till the last interceptor has finished for getting the call.

[1] https://github.com/grpc/grpc/pull/21455/files#diff-0894ffe42420c99925e97890b1f2685bR178